### PR TITLE
Implement EXTCODECOPY

### LINKED ISSUE #269: Implement extcodecopy
## Description
Currently the codecopy instruction for laser  is not implemented

### PR DESCRIPTION
**Pull request by norhh**
fixes #269

### CODE REVIEW

File: mythril/laser/ethereum/instructions.py [hunk 1]
```diff
@@ -944,16 +944,66 @@ def codecopy_(self, global_state: GlobalState) -> List[GlobalState]:
...
+        environment = global_state.environment
+        try:
+            addr = hex(helper.get_concrete_int(addr))
+        except TypeError:
+            log.debug("unsupported symbolic address for EXTCODESIZE")
+            state.stack.append(global_state.new_bitvec("extcodesize_" + str(addr), 256))
+            return [global_state]
+
+        try:
+            code = self.dynamic_loader.dynld(environment.active_account.address, addr)
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]: We should first check if the current environment already contains an account with the specific address

Comment (JoranHonig) [reply mythril/laser/ethereum/instructions.py]: The same would hold for extcodecopy

File: mythril/laser/ethereum/instructions.py [hunk 2]
```diff
@@ -984,14 +1045,13 @@ def codecopy_(self, global_state: GlobalState) -> List[GlobalState]:
                 )
             return [global_state]
```
Comment (nbanmp) [inline mythril/laser/ethereum/instructions.py]: Very minor note: The return value in `_code_copy_helper` is unused. We can do `return self._code_copy_helper(...)` in `codecopy` and `extcodecopy` or we can delete the return values in `_code_copy_helper` and save a line or two of code. Or we can just leave things as is.

File: mythril/laser/ethereum/instructions.py [hunk 3]
```diff
@@ -944,16 +944,77 @@ def codecopy_(self, global_state: GlobalState) -> List[GlobalState]:
...
+            code_offset=code_offset,
+            size=size,
+            op="CODECOPY",
+            global_state=global_state,
+        )
+
+        return [global_state]
+
+    def _get_code_from_address(self, global_state: GlobalState, addr: str) -> str:
+        if addr in global_state.accounts:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]: Hmm, not 100% sure but shouldn't it be:
```suggestion
        if addr in global_state.accounts.keys():
```

Comment (JoranHonig) [reply mythril/laser/ethereum/instructions.py]: Maybe this is implicitly done by python here

File: mythril/laser/ethereum/instructions.py [hunk 4]
```diff
@@ -944,16 +944,77 @@ def codecopy_(self, global_state: GlobalState) -> List[GlobalState]:
...
+            memory_offset=memory_offset,
+            code_offset=code_offset,
+            size=size,
+            op="CODECOPY",
+            global_state=global_state,
+        )
+
+        return [global_state]
+
+    def _get_code_from_address(self, global_state: GlobalState, addr: str) -> str:
```
Comment (JoranHonig) [inline mythril/laser/ethereum/instructions.py]: Could you create some unit tests for this function

Comment (JoranHonig) [reply mythril/laser/ethereum/instructions.py]: cover the two different edge cases 👍

**Comment by JoranHonig** [comment]
Looking good 👍

File: mythril/laser/ethereum/instructions.py (hunk 3 above)
Comment (norhh) [reply mythril/laser/ethereum/instructions.py]: Yep

File: mythril/laser/ethereum/instructions.py [hunk 5]
```diff
@@ -1093,35 +1137,41 @@ def codecopy_(self, global_state: GlobalState) -> List[GlobalState]:
...
+    def extcodecopy_(self, global_state: GlobalState) -> List[GlobalState]:
         """
 
         :param global_state:
         :return:
         """
         state = global_state.mstate
         addr = state.stack.pop()
-        environment = global_state.environment
+        memory_offset, code_offset, size = (
```
Comment (nbanmp) [inline mythril/laser/ethereum/instructions.py]: This can be:
```
addr, memory_offset, code_offset, size = (
   <pops>
)
```
